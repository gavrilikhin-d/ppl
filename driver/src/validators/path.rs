use std::path::PathBuf;

use miette::{Diagnostic, SourceSpan};
use thiserror::Error;

/// Diagnostic for a wrong path
#[derive(Error, Debug, Diagnostic)]
#[error("path doesn't exist")]
#[diagnostic(code(driver::wrong_path))]
pub struct WrongPath {
    /// Span to the path
    #[label("path doesn't exist")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_path`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingPathError {
    #[error(transparent)]
    #[diagnostic(code(driver::io_error))]
    IoError(#[from] std::io::Error),

    #[error(transparent)]
    WrongPath(#[from] WrongPath),
}

/// Check that path exists
pub fn exists(path: &str) -> Result<PathBuf, ExistingPathError> {
    let path = PathBuf::from(path);
    if path.try_exists()? {
        return Ok(path);
    }
    Err(WrongPath { at: (0..0).into() }.into())
}

/// Diagnostic for a path, that's not a directory
#[derive(Error, Debug, Diagnostic)]
#[error("expected path to directory")]
#[diagnostic(code(driver::expected_directory))]
pub struct ExpectedDirectory {
    /// Span to the path
    #[label("isn't a directory")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_dir`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingDirError {
    #[error(transparent)]
    ExistingPathError(#[from] ExistingPathError),

    #[error(transparent)]
    ExpectedDirectory(#[from] ExpectedDirectory),
}

/// Check that path exists and is a directory
pub fn is_dir(path: &str) -> Result<PathBuf, ExistingDirError> {
    let p = exists(path)?;
    if p.is_dir() {
        return Ok(p);
    }
    Err(ExpectedDirectory { at: (0, 0).into() }.into())
}
