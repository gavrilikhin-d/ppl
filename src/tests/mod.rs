use crate::e2es;

e2es! {
    array,
    candidate_not_viable,
    clone,
    common_functions,
    constraints,
    constraints_in_constructor,
    consume_greater,
    deps,
    deref_member_ref,
    destructor,
    empty_block,
    empty_constructor,
    escaped_id,
    generics,
    import_all,
    integer,
    integer_not_eq_rational,
    invalid_indentation,
    memory,
    missing_fields,
    monomorphize,
    monomorphize_predeclared,
    multifile,
    multiple_errors,
    multiple_initialization,
    non_class_constructor,
    plus_assign,
    predeclare_function,
    predeclare_vars,
    rational,
    reference_mut,
    reference_to_none,
    references,
    specify_variable_ty,
    star,
    string,
    supertraits,
    traits,
    type_as_value,
    trait_with_ref,
    type_of,
    wrong_initializer_type
}

#[test]
fn ppl() {
    use std::path::Path;

    use insta::assert_snapshot;
    use tempdir::TempDir;

    // Compile-time check that file exists
    include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl/src/lib.ppl"));

    let temp_dir = TempDir::new("ppl").unwrap();
    let tmp = temp_dir.path();
    let name = "ppl";
    let dir = Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl"));

    let res = crate::e2e::internal::compile(&tmp, dir);
    if let Err(err) = res {
        assert_snapshot!("ppl.error", err);
        return;
    }

    let hir = crate::e2e::internal::hir(&tmp, name, &dir);
    assert_snapshot!("ppl.hir", hir);

    let ir = crate::e2e::internal::ir(&tmp, name, &dir);
    assert_snapshot!("ppl.ir", ir);
}
