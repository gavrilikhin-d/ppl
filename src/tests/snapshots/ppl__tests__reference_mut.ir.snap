---
source: src/tests/mod.rs
expression: ir
---
; ModuleID = 'main'
source_filename = "src/main.ppl"

%"Type<String>" = type { %String, %Integer }
%String = type { ptr }
%Integer = type { ptr }

@"Type<String>" = private global %"Type<String>" zeroinitializer
@0 = private unnamed_addr constant [7 x i8] c"String\00", align 1
@x = global %Integer zeroinitializer
@"$tmp0@43" = global ptr null
@y = global ptr null
@"$tmp0@53" = global %Integer zeroinitializer
@"$tmp0@69" = global %Integer zeroinitializer
@"$tmp0@79" = global %Integer zeroinitializer

define private void @initialize() !dbg !3 {
  %1 = alloca %"Type<String>", align 8, !dbg !7
  %"Type<String>.name" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 0, !dbg !7
  %2 = call %String @string_from_c_string_and_length(ptr @0, i64 6), !dbg !8
  store %String %2, ptr %"Type<String>.name", align 8, !dbg !8
  %"Type<String>.size" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 1, !dbg !8
  %3 = call %Integer @integer_from_i64(i64 8), !dbg !8
  store %Integer %3, ptr %"Type<String>.size", align 8, !dbg !8
  %4 = load %"Type<String>", ptr %1, align 8, !dbg !8
  store %"Type<String>" %4, ptr @"Type<String>", align 8, !dbg !8
  br label %return, !dbg !8

return:                                           ; preds = %0
  ret void
}

declare %String @string_from_c_string_and_length(ptr, i64)

declare %Integer @integer_from_i64(i64)

define private void @initialize.1() !dbg !9 {
  %1 = call %Integer @integer_from_i64(i64 1), !dbg !10
  store %Integer %1, ptr @x, align 8, !dbg !10
  br label %return, !dbg !10

return:                                           ; preds = %0
  ret void
}

define private void @initialize.2() !dbg !11 {
  store ptr @x, ptr @"$tmp0@43", align 8, !dbg !12
  br label %return, !dbg !12

return:                                           ; preds = %0
  ret void
}

define private void @initialize.3() !dbg !13 {
  %1 = call ptr @"reference to mutable <:ReferenceMut<Integer>>"(ptr @"$tmp0@43"), !dbg !14
  store ptr %1, ptr @y, align 8, !dbg !14
  br label %return, !dbg !14

return:                                           ; preds = %0
  ret void
}

define private ptr @"reference to mutable <:ReferenceMut<Integer>>"(ptr %0) !dbg !15 {
  %return_value = alloca ptr, align 8
  %ref = alloca ptr, align 8
  store ptr %0, ptr %ref, align 8
  %2 = load ptr, ptr %ref, align 8, !dbg !16
  store ptr %2, ptr %return_value, align 8, !dbg !16
  br label %return, !dbg !16

return:                                           ; preds = %1
  %3 = load ptr, ptr %return_value, align 8
  ret ptr %3
}

define private void @initialize.4() !dbg !17 {
  %1 = load ptr, ptr @y, align 8, !dbg !18
  %2 = load %Integer, ptr %1, align 8, !dbg !18
  store %Integer %2, ptr @"$tmp0@53", align 8, !dbg !18
  br label %return, !dbg !18

return:                                           ; preds = %0
  ret void
}

define private void @initialize.5() !dbg !19 {
  %1 = load ptr, ptr @y, align 8, !dbg !20
  %2 = load %Integer, ptr %1, align 8, !dbg !20
  store %Integer %2, ptr @"$tmp0@69", align 8, !dbg !20
  br label %return, !dbg !20

return:                                           ; preds = %0
  ret void
}

define private void @initialize.6() !dbg !21 {
  %1 = load %Integer, ptr @x, align 8, !dbg !22
  %2 = call %Integer @clone_integer(%Integer %1), !dbg !22
  store %Integer %2, ptr @"$tmp0@79", align 8, !dbg !22
  br label %return, !dbg !22

return:                                           ; preds = %0
  ret void
}

declare %Integer @clone_integer(%Integer)

define void @main.execute() !dbg !23 {
init_globals:
  call void @initialize(), !dbg !24
  call void @initialize.1(), !dbg !25
  call void @initialize.2(), !dbg !26
  call void @initialize.3(), !dbg !27
  call void @initialize.4(), !dbg !28
  call void @initialize.5(), !dbg !29
  call void @initialize.6(), !dbg !30
  br label %0, !dbg !30

0:                                                ; preds = %init_globals
  %1 = load ptr, ptr @y, align 8, !dbg !31
  %2 = call %Integer @integer_from_i64(i64 2), !dbg !32
  store %Integer %2, ptr %1, align 8, !dbg !32
  %3 = load %Integer, ptr @x, align 8, !dbg !33
  call void @destroy_integer(%Integer %3), !dbg !33
  %4 = load %Integer, ptr @"$tmp0@53", align 8, !dbg !28
  call void @destroy_integer(%Integer %4), !dbg !28
  %5 = load %Integer, ptr @"$tmp0@69", align 8, !dbg !29
  call void @destroy_integer(%Integer %5), !dbg !29
  %6 = load %Integer, ptr @"$tmp0@79", align 8, !dbg !30
  call void @destroy_integer(%Integer %6), !dbg !30
  br label %return, !dbg !30

return:                                           ; preds = %0
  ret void
}

declare void @destroy_integer(%Integer)

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "ppl", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false, sysroot: "/")
!2 = !DIFile(filename: "src/main.ppl", directory: ".")
!3 = distinct !DISubprogram(name: "initialize", linkageName: "initialize", scope: !2, file: !2, line: 6, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DIBasicType(name: "i32", size: 32, encoding: DW_ATE_signed)
!7 = !DILocation(line: 6, scope: !3)
!8 = !DILocation(line: 0, scope: !3)
!9 = distinct !DISubprogram(name: "initialize.1", linkageName: "initialize.1", scope: !2, file: !2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!10 = !DILocation(line: 0, column: 12, scope: !9)
!11 = distinct !DISubprogram(name: "initialize.2", linkageName: "initialize.2", scope: !2, file: !2, line: 1, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!12 = !DILocation(line: 1, column: 29, scope: !11)
!13 = distinct !DISubprogram(name: "initialize.3", linkageName: "initialize.3", scope: !2, file: !2, line: 1, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!14 = !DILocation(line: 1, column: 29, scope: !13)
!15 = distinct !DISubprogram(name: "reference to mutable <:ReferenceMut<Integer>>", linkageName: "reference to mutable <:ReferenceMut<Integer>>", scope: !13, file: !2, line: 6, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!16 = !DILocation(line: 6, scope: !15)
!17 = distinct !DISubprogram(name: "initialize.4", linkageName: "initialize.4", scope: !2, file: !2, line: 2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!18 = !DILocation(line: 2, column: 8, scope: !17)
!19 = distinct !DISubprogram(name: "initialize.5", linkageName: "initialize.5", scope: !2, file: !2, line: 4, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!20 = !DILocation(line: 4, column: 8, scope: !19)
!21 = distinct !DISubprogram(name: "initialize.6", linkageName: "initialize.6", scope: !2, file: !2, line: 5, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!22 = !DILocation(line: 5, column: 8, scope: !21)
!23 = distinct !DISubprogram(name: "main.execute", linkageName: "main.execute", scope: !2, file: !2, line: 2, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!24 = !DILocation(line: 6, scope: !23)
!25 = !DILocation(line: 0, column: 12, scope: !23)
!26 = !DILocation(line: 1, column: 29, scope: !23)
!27 = !DILocation(line: 1, column: 8, scope: !23)
!28 = !DILocation(line: 2, column: 8, scope: !23)
!29 = !DILocation(line: 4, column: 8, scope: !23)
!30 = !DILocation(line: 5, column: 8, scope: !23)
!31 = !DILocation(line: 3, scope: !23)
!32 = !DILocation(line: 3, column: 4, scope: !23)
!33 = !DILocation(line: 0, scope: !23)
